src/rust/runtime/memory/mod.rs

```diff
@@ -29,143 +29,130 @@ pub use self::{buffer_pool::*, demibuffer::*};
+    let buf: DemiBuffer = unsafe { DemiBuffer::from_raw(token) };
+    drop(buf);
 
-        if sga.sga_buf == ptr::null_mut() {
-            return Err(Fail::new(libc::EINVAL, "demi_sgarray_t has invalid DemiBuffer token"));
-        }
+    Ok(())
+}
 
-        // Convert back to a DemiBuffer.
-        // Safety: The `NonNull::new_unchecked()` call is safe, as we verified `sga.sga_buf` is not null above.
-        let token: NonNull<u8> = unsafe { NonNull::new_unchecked(sga.sga_buf as *mut u8) };
-        // Safety: The `DemiBuffer::from_raw()` call *should* be safe, as the `sga_buf` field in the `demi_sgarray_t`
-        // contained a valid `DemiBuffer` token when we provided it to the user (and the user shouldn't change it).
-        let buf: DemiBuffer = unsafe { DemiBuffer::from_raw(token) };
-        let mut clone: DemiBuffer = buf.clone();
-
-        // Don't drop buf, as it holds the same reference to the data as the sgarray (which should keep it).
-        mem::forget(buf);
-
-        // Check to see if the user has reduced the size of the buffer described by the sgarray segment since we
-        // provided it to them.  They could have increased the starting address of the buffer (`sgaseg_buf`),
-        // decreased the ending address of the buffer (`sgaseg_buf + sgaseg_len`), or both.
-        let sga_data: *const u8 = sga.sga_segs[0].sgaseg_buf as *const u8;
-        let sga_len: usize = sga.sga_segs[0].sgaseg_len as usize;
-        let clone_data: *const u8 = clone.as_ptr();
-        let mut clone_len: usize = clone.len();
-        if sga_data != clone_data || sga_len != clone_len {
-            // We need to adjust the DemiBuffer to match the user's changes.
-
-            // First check that the user didn't do something non-sensical, like change the buffer description to
-            // reference address space outside of the DemiBuffer's allocated memory area.
-            if sga_data < clone_data || sga_data.addr() + sga_len > clone_data.addr() + clone_len {
-                return Err(Fail::new(
-                    libc::EINVAL,
-                    "demi_sgarray_t describes data outside backing buffer's allocated region",
-                ));
-            }
-
-            // Calculate the amount the new starting address is ahead of the old.  And then adjust `clone` to match.
-            let adjustment_amount: usize = sga_data.addr() - clone_data.addr();
-            clone.adjust(adjustment_amount)?;
-
-            // An adjustment above would have reduced clone.len() by the adjustment amount.
-            clone_len -= adjustment_amount;
-            debug_assert_eq!(clone_len, clone.len());
-
-            // Trim the clone down to size.
-            let trim_amount: usize = clone_len - sga_len;
-            clone.trim(trim_amount)?;
+/// Clones a scatter-gather array.
+pub fn clone_sgarray(sga: &demi_sgarray_t) -> Result<DemiBuffer, Fail> {
```

Comment:
This fn is long, can it be broken into smaller chunks? The mechanisms to clone and the checks can be separated into distinct fns.